### PR TITLE
Add ReduceAllGPU kernel

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
- It adds new feature needed because we need to have max reduction as part of ToDecibels implementation

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Added ReduceAllGPU kernel*
 - Affected modules and functionalities:
     *ReduceAllGPU kernel*
 - Key points relevant for the review:
     *The kernel implementation*
 - Validation and testing:
     *C++ tests added*
 - Documentation (including examples):
     *N/A*

**JIRA TASK**: *[DALI-1345]*

### CODE REVIEW

File: dali/kernels/reduce/reduce_all_kernel_gpu.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                    cudaMemcpyHostToDevice, context.gpu.stream);
+
+    auto* buffer_gpu = context.scratchpad->Allocate<T>(AllocType::GPU, tmp_buffer_size_);
+
+    dim3 block(32, 32);
+    dim3 grid(blocks_per_sample_, num_samples);
+    ReduceAllBatchedKernel<<<grid, block>>>(buffer_gpu, sample_data_gpu, sample_size_gpu,
+                                            reduction);
+
+    DALI_ENFORCE(out.is_contiguous(), "Reduce all kernel expects the output to be contiguous");
```
Comment (mzient) [inline dali/kernels/reduce/reduce_all_kernel_gpu.h]: Move this test to the top - there's no point starting work when this is not satisfied.

File: dali/kernels/reduce/reduce_all_kernel_gpu.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    auto* sample_size = context.scratchpad->Allocate<int64_t>(AllocType::Host, num_samples);
+    for (int i = 0; i < num_samples; i++) {
+      sample_data[i] = in.tensor_data(i);
+      sample_size[i] = volume(in.tensor_shape(i));
+    }
+
+    auto* sample_data_gpu = context.scratchpad->Allocate<const T*>(AllocType::GPU, num_samples);
+    auto* sample_size_gpu = context.scratchpad->Allocate<int64_t>(AllocType::GPU, num_samples);
+    // Single memcpy, since the data in the scratchpad is contiguous
+    cudaMemcpyAsync(sample_data_gpu, sample_data,
```
Comment (mzient) [inline dali/kernels/reduce/reduce_all_kernel_gpu.h]: CUDA_CALL

File: dali/kernels/reduce/reduce_all_kernel_gpu.h [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+#include "dali/core/format.h"
+#include "dali/core/util.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/reduce/reduce.h"
+
+namespace dali {
+namespace kernels {
+namespace reduce {
+
+template <typename T, int ndim, typename Reductor, typename Preprocessor = identity>
```
Comment (mzient) [inline dali/kernels/reduce/reduce_all_kernel_gpu.h]:
```suggestion
template <typename T, int ndim, typename Reducer, typename Preprocessor = identity>
```

File: dali/kernels/reduce/reduce_all_kernel_gpu.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  DLL_PUBLIC KernelRequirements Setup(KernelContext &context,
+                                      const InListGPU<T, ndim> &in) {
+    auto num_samples = in.size();
+    ScratchpadEstimator se;
+
+    int64_t max_sample_size = 0;
+    for (int i = 0; i < num_samples; i++) {
+      max_sample_size = std::max(max_sample_size, volume(in.tensor_shape(i)));
+    }
+    blocks_per_sample_ = std::max(32, static_cast<int>(std::sqrt(max_sample_size)));
```
Comment (mzient) [inline dali/kernels/reduce/reduce_all_kernel_gpu.h]: I'm starting to have second thoughts here. For small data (on ther order of < 4kB max) there's no point in splitting the reduction into two stages - the kernel invocation would constitute significant overhead. For such a case, you can force blocks_per_sample to 1 and make the first kernel produce the output directly.

File: dali/kernels/reduce/reduce_all_kernel_gpu.h [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    // Single memcpy, since the data in the scratchpad is contiguous
+    cudaMemcpyAsync(sample_data_gpu, sample_data,
+                    num_samples * sizeof(const T*) + num_samples * sizeof(int64_t),
+                    cudaMemcpyHostToDevice, context.gpu.stream);
+
+    auto* buffer_gpu = context.scratchpad->Allocate<T>(AllocType::GPU, tmp_buffer_size_);
+
+    dim3 block(32, 32);
+    dim3 grid(blocks_per_sample_, num_samples);
+    ReduceAllBatchedKernel<<<grid, block>>>(buffer_gpu, sample_data_gpu, sample_size_gpu,
```
Comment (mzient) [inline dali/kernels/reduce/reduce_all_kernel_gpu.h]:
```suggestion
    ReduceAllBatchedKernel<<<grid, block, 0, context.gpu.stream>>>(buffer_gpu, sample_data_gpu, sample_size_gpu,
```

File: dali/kernels/reduce/reduce_all_kernel_gpu.h [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                                            reduction);
+
+    DALI_ENFORCE(out.is_contiguous(), "Reduce all kernel expects the output to be contiguous");
+    auto* out_start = out[0].data;
+
+    dim3 grid2(1, num_samples);
+    ReduceAllBlockwiseKernel<<<grid2, block>>>(out_start, buffer_gpu, blocks_per_sample_,
```
Comment (mzient) [inline dali/kernels/reduce/reduce_all_kernel_gpu.h]:
```suggestion
    ReduceAllBlockwiseKernel<<<grid2, block, 0, context.gpu.stream>>>(out_start, buffer_gpu, blocks_per_sample_,
```

Comment (jantonguirao) [reply dali/kernels/reduce/reduce_all_kernel_gpu.h]: good catch

File: dali/kernels/reduce/reduce_all_kernel_gpu.h (hunk 4 above)
Comment (jantonguirao) [reply dali/kernels/reduce/reduce_all_kernel_gpu.h]: done

File: dali/kernels/reduce/reduce_all_kernel_gpu.h (hunk 3 above)
Comment (jantonguirao) [reply dali/kernels/reduce/reduce_all_kernel_gpu.h]: done

File: dali/kernels/reduce/reduce_all_kernel_gpu.h (hunk 2 above)
Comment (jantonguirao) [reply dali/kernels/reduce/reduce_all_kernel_gpu.h]: done

File: dali/kernels/reduce/reduce_all_kernel_gpu.h (hunk 1 above)
Comment (jantonguirao) [reply dali/kernels/reduce/reduce_all_kernel_gpu.h]: done

**Pull request merged**